### PR TITLE
Migrate to Bzlmod

### PR DESCRIPTION
**Pull request by meteorcloudy**
- Migrated all Bazel dependencies to Bazel modules
- Manage Go dependencies with rules_go + go.mod
- Manage Python dependencies with rules_python + requirements files
- Manage Java dependencies with rules_jvm_external
- Depend on a development version of rules_scala and uses the new setup for scala_test
- Implemented a module extension for introducing browsers dependencies.

### CODE REVIEW

File: web/versioned/browsers-0.3.1.bzl
Comment (gkdn) [inline web/versioned/browsers-0.3.1.bzl:208]: I was trying to figure this out. Does it need to be configured? Or could we enable all without consequences? (e.g. slower loading/builds, more downloads etc.)

File: MODULE.bazel
Comment (meteorcloudy) [reply MODULE.bazel:99]: With Bzlmod, dependencies now work transitively. Every module just needs to declare its direct dependencies. So no need for your users to load dependencies through helper function anymore.

If you have dependencies that's only used by tests and don't want them to be propagated to your dependents, you can set `dev_dependency = True` on `bazel_dep` or `use_extension` directive. 

See https://bazel.build/rules/lib/globals/module#bazel_dep

File: MODULE.bazel
Comment (gkdn) [reply MODULE.bazel]: > Did you have specific reason for this?

This is both velocity and practicality problem:

Many times we see problems during a release and patch other projects to do the release. With this we will be required to wait any fixes to be released before doing our own release. Waiting for a release for things that we don't have control on will be a huge velocity concern (even if we automate/steamline BCR process for our own repos).

Similarly sometimes to avoid larger migration and compatibility issues, we use a commit between official releases that avoid the problem. That will also start being a problem.

Another problem that exaggerates the issue is, we have multiple repos that work on the head. Also we have sub-workspaces to test out the setup as seen from outside of our workspace. And I also expect some users to use our head version (which was common in the past since we don't do frequent releases). When we need to rely on a patch like this for any reason, it will need to be copied over every single workspace where in most cases they don't even know about J2CL dependencies.

> Because dependency resolution won't work 

Would it be possible to declare closest version where the commit is based on? 
So it will be like `N + epsilon` where it will be newer than version `N`, older than `N+1` but not comparable with any other `N + epsilon`s. In a case where it is not comparable the users needs to resolve by override as before. That could significantly reduce the need for user overrides.

File: MODULE.bazel
Comment (gkdn) [reply MODULE.bazel:99]: Nice. Are "dev_dependency"s only improve the Bazel while loading the graph or it is actually required to reduce what get downloaded? (again asking to improve my mental model)

File: web/versioned/browsers-0.3.1.bzl
Comment (gkdn) [reply web/versioned/browsers-0.3.1.bzl:208]: In case you missed this one.

**Comment by gkdn** [comment]
LGTM. Adding @mtrea.
(@mtrea There are a couple of discussions but mostly educational)

**Review by gkdn — APPROVED** [review]

File: MODULE.bazel
Comment (meteorcloudy) [reply MODULE.bazel:100]: dev_dependency from non root modules will not show up in the transitive dependency graph, so it can never be downloaded.

File: web/versioned/browsers-0.3.1.bzl
Comment (meteorcloudy) [reply web/versioned/browsers-0.3.1.bzl]: Yeah, I actually wants to implement this properly to show case how module extension works. Please see [this commit](https://github.com/bazelbuild/rules_webtesting/pull/478/commits/ecf708ee013a25ad7ffe8c896f4653cf8ad93576#diff-8f23351c622c1fa4118fe0289a67ee47c2445c8b7278f9e9ad2cf619573327dc).

It basically offers:
- a `version` tag to be used by any module to specify the version of browsers version. If multiple versions are specified, the highest one is chosen. (e.g module A requires 0.3.3, module B requires 0.3.4, then 0.3.4 is used eventually)
- a `override_version` tag to be used by the **root** module to pin the version. If a non-module root uses this tag, it'll be ignored.

Overall, this implements a reasonable resolution logic when there are version conflicts while also gives the root module full control when needed.

In addition, I removed all `firefox = False, chromium = False, sauce = False` parameters, because it's fine to declare all those repos, they will be fetched/downloaded lazily when the targets you build actually need them.

Let me know if this make sense for rules_webtesting.

File: MODULE.bazel
Comment (meteorcloudy) [reply MODULE.bazel]: > Waiting for a release for things that we don't have control on will be a huge velocity concern

One thing we could do is to upstream the fix as a [BCR patch version](https://github.com/bazelbuild/bazel-central-registry/blob/main/docs/README.md#versions-format) if the module maintainer is too slow. Anyway can send a PR and a registry maintainer can approve. But in general, I think at least for popular Bazel rules, they are well supported in BCR.

> we have multiple repos that work on the head. Also we have sub-workspaces to test out the setup as seen from outside of our workspace.

If they are living in the same repo, then they could probably share the same overrides via [include](https://bazel.build/rules/lib/globals/module#include) in MODULE.bazel with some symlink tricks?

> Would it be possible to declare closest version where the commit is based on?

This is big feature request that needs a lot of thoughts, I'm not sure it's a good idea to do that. @Wyverald 

I'm happy to look at some specific cases where you have to use commits instead a released version and see if we can find a better process or should actually change Bzlmod to support it better.

File: web/versioned/browsers-0.3.1.bzl
Comment (gkdn) [reply web/versioned/browsers-0.3.1.bzl]: Looks like it matches the current functionality and provide in a nice way to handle conflict so +1.

I was wondering if it would be easy to allow the module to ask to alias a specific version so multiple could be used in parallel without conflict.

For example something like:
```
browser_repositories = use_extension("//web:extension.bzl", "browser_repositories_extension")
browser_repositories.install(version = "0.3.3", alias = "my_browsers")
use_repo(browser_repositories, "my_browsers")
```

Then user can do something like `@my_browsers//:chrome` without causing conflict.

Though it seems like that requires a much larger refactoring.

**Comment by mtrea** [comment]
At a high level here, what are you trying to accomplish? Looking at the top-level description:

>Migrated all Bazel dependencies to Bazel modules

What's wrong with Bazel dependencies? Are they deprecated and Bazel modules are the new thing to use now?

>Manage Go dependencies with rules_go + go.mod
>Manage Python dependencies with rules_python + requirements files
>Manage Java dependencies with rules_jvm_external

Why? Is this like a tech debt thing, or does it add new capabilities?

>Depend on a development version of rules_scala and uses the new setup for scala_test

Why is this preferred over the old way of getting rules_scala? Is it because "Otherwise, the root module has to declare the same override"? (to quote your change in MODULE.bazel)

>Implemented a module extension for introducing browsers dependencies.

Are you planning to use this module extension in a followup?

**Comment by gkdn** [comment]
> What's wrong with Bazel dependencies? Are they deprecated and Bazel modules are the new thing to use now?

Bazel has deprecated old WORKSPACE style dependency management with Bzlmod. With Bazel 8, existing approach will be disable. So this PR replaces old style with the modern Bzlmod system.
More info: https://bazel.build/external/migration

File: MODULE.bazel
Comment (meteorcloudy) [reply MODULE.bazel:91]: For just overriding the version, no. If another module also wants to reference those repos in labels, then it must add `use_repo` to make the repo visible.

**Comment by meteorcloudy** [comment]
Yes, it's one of the most significant changes since Bazel was open sourced. You can also find a lot of Bzlmod talks online: https://www.youtube.com/results?search_query=bzlmod

File: web/extension.bzl
Comment (meteorcloudy) [reply web/extension.bzl]: Sounds good, done!

File: web/versioned/browsers-0.3.1.bzl
Comment (meteorcloudy) [reply web/versioned/browsers-0.3.1.bzl]: > I was wondering if it would be easy to allow the module to ask to alias a specific version so multiple could be used in parallel without conflict.

So this module extension is mostly used to configure which version of browser repositories will be used by rules_webtesting. If other modules wants to use a specific version, they can still write a new module extension to call `browser_repositories` from a specific versioned bzl file like `web/versioned/browsers-0.3.1.bzl`, and it will be their "private" dependencies.
You can also call repository rule directly in MODULE.bazel via [`use_repo_rule`](https://bazel.build/rules/lib/repo/http#setup).

**Comment by mtrea** [comment]
Had a chat with Goktug and he clarified that everything comes back to this Bzlmod topic, even e.g. the rules_scala change, which needs to use a bzlmod compatible version of scala. It seems like this is a pure modernization of how we do the Bazel stuff, and it doesn't have any internal impact, so this seems reasonable to me.

**Review by mtrea — APPROVED** [review]

**Pull request merged**
